dotcom-rendering/src/web/browser/islands/initHydration.ts

```diff
@@ -27,18 +28,28 @@ export const initHydration = (elements: NodeListOf<Element>): void => {
 			switch (deferUntil) {
 				case 'idle': {
 					whenIdle(() => {
-						doHydration(name, props, element);
+						void doHydration(name, props, element);
 					});
 					break;
 				}
 				case 'visible': {
 					whenVisible(element, () => {
-						doHydration(name, props, element);
+						void doHydration(name, props, element);
+					});
+					break;
+				}
+				case 'interaction': {
+					onInteraction(element, (targetElement) => {
+						void doHydration(name, props, element).then(() => {
+							targetElement.dispatchEvent(
+								new MouseEvent('click'),
```

Comment:
this is a great idea!

I'm wondering how the replay functionality will work in the case where the  placeholder is clicked more than once before the hydration is complete?

1. Will it attempt to hydrate again? I guess SWR will be caching the fetch call, but afaict `doHydration` might end up running again because the `guReady` flag is only set after the hydration is complete.
  - We wouldn't want to prevent retries entirely, in the case where it fails, though, so perhaps in practice a simple 'throttling' function would do the trick?
2. If the user clicks more than once before the hydration is complete, will all of the clicks be stored to be replayed after hydration? (E.g. in the counter case we'd plausibly want to replay multiple clicks, though that's debatable.)

Reply:
Great point!

To my understanding any clicks made in the dead zone (aka the uncanny valley) after the initial one that triggers hydration and before the component has been fully hydrated will be lost. Lost clicks are an accepted part of the hydration problem, even if you hydrate the whole page or give something the highest priority for hydration, there's always a gap where interaction gets lost.

In our case I think the trade off we're making is worth it. Deferring until click does increase the risk there will be interaction during the dead zone but I think the fact we capture and replay the first event means we probably offer a better experience for most readers.

Reply:
The current functionality is quite primitive, the event listener will only run once, which triggers the hydration, and then the event replays. As Oliver says, any other clicks will be lost - but we don't risk trying to hydrate it multiple times.

Right now the 'decoupled' nature of the the 'trigger' and the actual hydration process means that a more advanced system would require more work, but perhaps that's what we need?

Reply:
I would focus on what we need right now, not what we might need in the future. Islands load rather quickly, especially if they are cached!